giantswarm/gsctl
# PoC: List clusters using Cluster API

Towards https://github.com/giantswarm/giantswarm/issues/9352

Please note that I'm asking for your review in order to check if I'm heading in the right direction. This is not ready to be merged, as it does not yet work with all installations.

### What's included?
* Logging in and authorization 
* Listing clusters (both V4 and V5)

### Disclaimer
* Currently only `ginger` is supported
  * because I've enforced the latest master version of `dex-app` there.
  * because there isn't any endpoint to get the installation CA certificate just yet

### How to test
In order to be able to use it as a `kubectl` plugin, the binary must be in a folder that is included in your `$PATH` variable, and the binary name should be prefixed with `kubectl-`. What comes after `kubectl-` is the `kubectl` command that will be used.

_Example:_
If we rename `kubectl-gs` to `kubectl-some-cool-command`, we would need to run `kubectl some cool command`.

#### Use as kubectl plugin
```shell
$ go build
$ cp ./gsctl /usr/local/bin/kubectl-gs
```
You now have a working `kubectl gs` plugin. (the app detects if it's running as a kubectl plugin)

#### Login 
```shell
$ kubectl gs login ginger.eu-west-1.aws.gigantic.io
```
(A browser window will open and you should give GitHub access to `dex`. After that, you should be logged in)

#### List clusters
```shell
$ kubectl gs list clusters
```

### What's left for this to be production-ready
* An installation info service: to be able to get the installation certificate, installation alias, provider etc.
* Tests
* Comments
* The oidc package functionality to be moved to `gscliauth`
* Multi-provider support?

I don't believe there's any documentation to be added just yet.

### Additional info
* The new OIDC implementation is supposed to replace the current one inside of `gscliauth` in the future, since it is much simpler, and is using the official Oauth2 package. While we still use Auth0, these 2 can co-exist together.
You can see some duplicated code between `gscliauth` and the new OIDC package. That was just for easier development. I will re-use the existing functionality in production.
* Check the comments below
